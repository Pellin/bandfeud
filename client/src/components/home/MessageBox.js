import React from 'react';

const MessageBox = ({ message }) => {
  return <div className="message-box">{message}</div>;
};

export default MessageBox;
