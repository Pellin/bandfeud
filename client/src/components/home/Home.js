import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { motion } from 'framer-motion';

import Header from '../Header';
import Play from '../home/play/Play';
import ShowBandlistModal from '../highscores/ShowBandlistModal';
import MessageBox from '../home/MessageBox';
import PlayIcon from '../../icons/PlayIcon';
import LogoSquare from '../../icons/LogoSquare';

import getHighscores from '../../actions/getHighscores';
import setMessage from '../../actions/setMessage';
import setOs from '../../actions/setOs';
import { gameOn } from '../../actions/gameStatus';

export const Home = ({
  bands,
  history,
  inGame,
  onSetMessage,
  onGetHighscores,
  onGameOn,
  onSetOs
}) => {
  useEffect(() => {
    onSetOs(navigator.userAgent);
    onGetHighscores();
  });
  const [buttonPressed, setButtonPressed] = useState(false);
  const [showFooter, setShowFooter] = useState(false);
  const [showModal, setShowModal] = useState(false);
  const [message, setMessage] = useState('');
  const startGame = () => {
    setButtonPressed(true);
    setMessage('Get ready...');
    setShowFooter(true);
    setTimeout(() => {
      setMessage('');
      onGameOn();
      onSetMessage('');
    }, 2000);
  };
  const releaseButton = () => {
    setButtonPressed(false);
  };
  const openModal = () => {
    setShowModal(true);
  };
  const closeModal = () => {
    setShowModal(false);
  };
  const buttonVariants = {
    notPressed: {
      opacity: 1,
      scale: 1,
      rotate: 720
    },
    pressed: {
      opacity: [null, 1, 0],
      scale: [null, 1.1, 0],
      rotate: [null, 780, 0]
    }
  };
  const logoVariants = {
    show: {
      opacity: 1
    },
    hide: {
      opacity: 0
    }
  }
  const footerVariants = {
    show: {
      y: 0
    },
    hide: {
      y: 100
    }
  };
  return (
    <div>
      <Header buttonPressed={buttonPressed} />
      <div>
        {!inGame && (
          <div className="home-container">
            <motion.div
              className="logo-square"
              initial={{ opacity: 0 }}
              animate={buttonPressed ? 'hide' : 'show'}
              transition={{ duration: 2 }}
              variants={logoVariants}
            >
              <LogoSquare />
            </motion.div>

            <div
              title="Start new game"
              className="new-game-button"
              onClick={startGame}
            >
              <motion.div
                className="play-button"
                initial={{ scale: 0, rotate: -360 }}
                animate={buttonPressed ? 'pressed' : 'notPressed'}
                transition={{ duration: 1.2, times: [0, 0.4, 1] }}
                variants={buttonVariants}
                onClick={startGame}
              >
                <PlayIcon />
              </motion.div>
            </div>
            {bands.length > 0 && !message ? (
              <div className="footer">
                <button className="show-modal-button" onClick={openModal}>
                  REVIEW LAST ROUND
                </button>
              </div>
            ) : (
              <div className="footer">
                <button className="show-modal-button-invisible">
                  NOT SHOWING
                </button>
              </div>
            )}
          </div>
        )}
      </div>
      {inGame && <Play releaseButton={releaseButton} history={history} />}
      <ShowBandlistModal
        bands={bands}
        closeModal={closeModal}
        showModal={showModal}
      />
      {message && (
        <motion.div
          className="footer"
          ease="ease-out"
          initial={{ y: 100 }}
          animate={showFooter ? 'show' : 'hide'}
          transition={{ duration: 0.3 }}
          variants={footerVariants}
        >
          {message && <MessageBox message={message} />}
        </motion.div>
      )}
    </div>
  );
};

const mapStateToProps = state => ({
  bands: state.bands,
  inGame: state.inGame
});

const mapDispatchToProps = dispatch => ({
  onSetMessage: message => dispatch(setMessage(message)),
  onGetHighscores: () => dispatch(getHighscores()),
  onGameOn: () => dispatch(gameOn()),
  onSetOs: os => dispatch(setOs(os))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Home);
