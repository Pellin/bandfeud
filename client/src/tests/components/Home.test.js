import React from 'react';
import { shallow } from 'enzyme';

import { Home } from '../../components/home/Home';

let onGetHighscoresSpy, onSetMessageSpy, onGameOnSpy, onSetOsSpy, bands;

beforeEach(() => {
  onGetHighscoresSpy = jest.fn();
  onSetMessageSpy = jest.fn();
  onGameOnSpy = jest.fn();
  onSetOsSpy = jest.fn();
  bands = [
    { name: 'one', url: 'www.1img.com' },
    { name: 'two', url: 'www.2img.com' },
    { name: 'three', url: 'www.3img.com' }
  ];
});

it('should render with PLAY <button> when inGame is false', () => {
  const wrapper = shallow(
    <Home
      bands={[]}
      inGame={false}
      onGetHighscores={onGetHighscoresSpy}
      onSetMessage={onSetMessageSpy}
      onSetOs={onSetOsSpy}
      onGameOn={onGameOnSpy}
      window={window}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

it('should render with PLAY <button> and Review... <button> when inGame is false and bands.length > 0', () => {
  const wrapper = shallow(
    <Home
      bands={bands}
      inGame={false}
      onGetHighscores={onGetHighscoresSpy}
      onSetMessage={onSetMessageSpy}
      onSetOs={onSetOsSpy}
      onGameOn={onGameOnSpy}
      window={window}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

it('should call onSetMessage and onGameOn correctly when "PLAY" button is clicked', () => {
  jest.useFakeTimers();
  const wrapper = shallow(
    <Home
      bands={[]}
      inGame={false}
      onGetHighscores={onGetHighscoresSpy}
      onSetMessage={onSetMessageSpy}
      onSetOs={onSetOsSpy}
      onGameOn={onGameOnSpy}
    />
  );

  wrapper.find('.new-game-button').simulate('click', { button: 0 });
  jest.runAllTimers();

  expect(onSetMessageSpy).toHaveBeenCalled();
  expect(onGameOnSpy).toHaveBeenCalled();

  jest.useRealTimers();
});

it('should render with <Play /> when inGame is true', () => {
  const wrapper = shallow(
    <Home
      inGame={true}
      onGetHighscores={onGetHighscoresSpy}
      onSetMessage={onSetMessageSpy}
      onSetOs={onSetOsSpy}
      onGameOn={onGameOnSpy}
    />
  );

  expect(wrapper).toMatchSnapshot();
});

it('should render ShowBandlistModal when "REVIEW LAST ROUND" button is clicked', () => {
  const wrapper = shallow(
    <Home
      bands={bands}
      inGame={false}
      onGetHighscores={onGetHighscoresSpy}
      onSetMessage={onSetMessageSpy}
      onSetOs={onSetOsSpy}
      onGameOn={onGameOnSpy}
    />
  );
  expect(wrapper.find('ShowBandlistModal').props().showModal).toBeFalsy();
  wrapper.find('button.show-modal-button').simulate('click', { button: 0 });
  expect(wrapper.find('ShowBandlistModal').props().showModal).toBeTruthy();
  expect(wrapper).toMatchSnapshot();
});
