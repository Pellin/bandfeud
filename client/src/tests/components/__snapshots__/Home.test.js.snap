// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ShowBandlistModal when "REVIEW LAST ROUND" button is clicked 1`] = `
<div>
  <Connect(Header)
    buttonPressed={false}
  />
  <div>
    <div
      className="home-container"
    >
      <ForwardRef(MotionComponent)
        animate="show"
        className="logo-square"
        initial={
          Object {
            "opacity": 0,
          }
        }
        transition={
          Object {
            "duration": 2,
          }
        }
        variants={
          Object {
            "hide": Object {
              "opacity": 0,
            },
            "show": Object {
              "opacity": 1,
            },
          }
        }
      >
        <SvgLogoSquare />
      </ForwardRef(MotionComponent)>
      <div
        className="new-game-button"
        onClick={[Function]}
        title="Start new game"
      >
        <ForwardRef(MotionComponent)
          animate="notPressed"
          className="play-button"
          initial={
            Object {
              "rotate": -360,
              "scale": 0,
            }
          }
          onClick={[Function]}
          transition={
            Object {
              "duration": 1.2,
              "times": Array [
                0,
                0.4,
                1,
              ],
            }
          }
          variants={
            Object {
              "notPressed": Object {
                "opacity": 1,
                "rotate": 720,
                "scale": 1,
              },
              "pressed": Object {
                "opacity": Array [
                  null,
                  1,
                  0,
                ],
                "rotate": Array [
                  null,
                  780,
                  0,
                ],
                "scale": Array [
                  null,
                  1.1,
                  0,
                ],
              },
            }
          }
        >
          <SvgPlayIcon />
        </ForwardRef(MotionComponent)>
      </div>
      <div
        className="footer"
      >
        <button
          className="show-modal-button"
          onClick={[Function]}
        >
          REVIEW LAST ROUND
        </button>
      </div>
    </div>
  </div>
  <ShowBandlistModal
    bands={
      Array [
        Object {
          "name": "one",
          "url": "www.1img.com",
        },
        Object {
          "name": "two",
          "url": "www.2img.com",
        },
        Object {
          "name": "three",
          "url": "www.3img.com",
        },
      ]
    }
    closeModal={[Function]}
    lastRound={true}
    showModal={true}
  />
</div>
`;

exports[`should render with <Play /> when inGame is true 1`] = `
<div>
  <Connect(Header)
    buttonPressed={false}
  />
  <div />
  <Connect(Play)
    releaseButton={[Function]}
  />
  <ShowBandlistModal
    closeModal={[Function]}
    lastRound={true}
    showModal={false}
  />
</div>
`;

exports[`should render with PLAY <button> and Review... <button> when inGame is false and bands.length > 0 1`] = `
<div>
  <Connect(Header)
    buttonPressed={false}
  />
  <div>
    <div
      className="home-container"
    >
      <ForwardRef(MotionComponent)
        animate="show"
        className="logo-square"
        initial={
          Object {
            "opacity": 0,
          }
        }
        transition={
          Object {
            "duration": 2,
          }
        }
        variants={
          Object {
            "hide": Object {
              "opacity": 0,
            },
            "show": Object {
              "opacity": 1,
            },
          }
        }
      >
        <SvgLogoSquare />
      </ForwardRef(MotionComponent)>
      <div
        className="new-game-button"
        onClick={[Function]}
        title="Start new game"
      >
        <ForwardRef(MotionComponent)
          animate="notPressed"
          className="play-button"
          initial={
            Object {
              "rotate": -360,
              "scale": 0,
            }
          }
          onClick={[Function]}
          transition={
            Object {
              "duration": 1.2,
              "times": Array [
                0,
                0.4,
                1,
              ],
            }
          }
          variants={
            Object {
              "notPressed": Object {
                "opacity": 1,
                "rotate": 720,
                "scale": 1,
              },
              "pressed": Object {
                "opacity": Array [
                  null,
                  1,
                  0,
                ],
                "rotate": Array [
                  null,
                  780,
                  0,
                ],
                "scale": Array [
                  null,
                  1.1,
                  0,
                ],
              },
            }
          }
        >
          <SvgPlayIcon />
        </ForwardRef(MotionComponent)>
      </div>
      <div
        className="footer"
      >
        <button
          className="show-modal-button"
          onClick={[Function]}
        >
          REVIEW LAST ROUND
        </button>
      </div>
    </div>
  </div>
  <ShowBandlistModal
    bands={
      Array [
        Object {
          "name": "one",
          "url": "www.1img.com",
        },
        Object {
          "name": "two",
          "url": "www.2img.com",
        },
        Object {
          "name": "three",
          "url": "www.3img.com",
        },
      ]
    }
    closeModal={[Function]}
    lastRound={true}
    showModal={false}
  />
</div>
`;

exports[`should render with PLAY <button> when inGame is false 1`] = `
<div>
  <Connect(Header)
    buttonPressed={false}
  />
  <div>
    <div
      className="home-container"
    >
      <ForwardRef(MotionComponent)
        animate="show"
        className="logo-square"
        initial={
          Object {
            "opacity": 0,
          }
        }
        transition={
          Object {
            "duration": 2,
          }
        }
        variants={
          Object {
            "hide": Object {
              "opacity": 0,
            },
            "show": Object {
              "opacity": 1,
            },
          }
        }
      >
        <SvgLogoSquare />
      </ForwardRef(MotionComponent)>
      <div
        className="new-game-button"
        onClick={[Function]}
        title="Start new game"
      >
        <ForwardRef(MotionComponent)
          animate="notPressed"
          className="play-button"
          initial={
            Object {
              "rotate": -360,
              "scale": 0,
            }
          }
          onClick={[Function]}
          transition={
            Object {
              "duration": 1.2,
              "times": Array [
                0,
                0.4,
                1,
              ],
            }
          }
          variants={
            Object {
              "notPressed": Object {
                "opacity": 1,
                "rotate": 720,
                "scale": 1,
              },
              "pressed": Object {
                "opacity": Array [
                  null,
                  1,
                  0,
                ],
                "rotate": Array [
                  null,
                  780,
                  0,
                ],
                "scale": Array [
                  null,
                  1.1,
                  0,
                ],
              },
            }
          }
        >
          <SvgPlayIcon />
        </ForwardRef(MotionComponent)>
      </div>
      <div
        className="footer"
      >
        <button
          className="show-modal-button-invisible"
        >
          NOT SHOWING
        </button>
      </div>
    </div>
  </div>
  <ShowBandlistModal
    bands={Array []}
    closeModal={[Function]}
    lastRound={true}
    showModal={false}
  />
</div>
`;
